### ISSUE TITLE
Method call on constrained typevar emits invalid-argument-type

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#1503]
Consider the following example. ty currently emits two errors for the `t.method()` call because `t.method` creates a union of bound method objects where `C.method` and `D.method` are bound to `T@f`, respectively. Instead, attribute access should properly distribute across the union `C | D` and yield a union of bound method objects where `C.method` is bound to an instance of `C`, and `D.method` is bound to an instance of `D`:

```py
class C:
    def method(self) -> int:
        return 0

class D:
    def method(self) -> str:
        return ""

def f[T: (C, D)](t: T):
    # Argument to bound method `method` is incorrect: Expected `C`, found `T@f` (invalid-argument-type)
    # Argument to bound method `method` is incorrect: Expected `D`, found `T@f` (invalid-argument-type)
    reveal_type(t.method())
```
https://play.ty.dev/c102d2bd-de4d-497c-9144-9f45def4196a

Related: https://github.com/astral-sh/ty/issues/138

**Comment by Wizzerinus** [comment]
Another example which seems related to this one:

```py
from typing import TypeVar

Num = TypeVar("Num", int, float)

def add(x: Num, y: Num):
    return x + y  # ty emits an error here
```

**Comment by danielgafni** [comment]
Yeah this is super annoying. I'm using [Narwhals](https://narwhals-dev.github.io/narwhals/basics/dataframe/) and this bug basically kills the entire library. 

```py
from narwhals.typing import FrameT

def fn(
    frame: FrameT
) -> FrameT:
    return frame.with_columns(
        nw.lit("hello world").alias("foo")
    )
```

```
error[invalid-argument-type]: Argument to bound method `with_columns` is incorrect
 --> src/metaxy/example.py:6:12
  |
5 | def fn(frame: FrameT) -> FrameT:
6 |     return frame.with_columns(nw.lit("hello world").alias("foo"))
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Argument type `FrameT@fn` does not satisfy upper bound `DataFrame[DataFrameT@DataFrame]` of type variable `Self`
  |


error[invalid-argument-type]: Argument to bound method `with_columns` is incorrect
 --> src/metaxy/example.py:6:12
  |
5 | def fn(frame: FrameT) -> FrameT:
6 |     return frame.with_columns(nw.lit("hello world").alias("foo"))
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Argument type `FrameT@fn` does not satisfy upper bound `LazyFrame[LazyFrameT@LazyFrame]` of type variable `Self`
  |

```

I am using `ty==0.0.12`

---

P.S. `ty`'s output for this `ty check` call has been 3k lines long - is this expected? It's printing the entire Python module where the original source code is defined.

**Comment by MichaReiser** [comment]
> P.S. ty's output for this ty check call has been 3k lines long - is this expected? It's printing the entire Python module where the original source code is defined.

This is not expected. It would be great if this is tracked in a separte issue

**Comment by carljm** [comment]
Example from #4162 

```py
from typing import TypeVar

class DF:
    def sort(self) -> "DF":
        return self

class LF:
    def sort(self) -> "LF":
        return self

T = TypeVar("T", DF, LF, "DF | LF")

def reassign(x: T) -> T:
    x = x.sort()   # error[invalid-assignment]: `DF | LF` is not assignable to `T`
    return x

def pass_union(v: "DF | LF") -> None:
    reassign(v)    # error[invalid-argument-type]: `DF | LF` not accepted for `T`
```

**Comment by notatallshaw-gts** [comment]
If I'm understanding the new errors ty is giving me since 0.0.70 this also rejects property reads, under a different rule, so the same cause now has two symptoms:

```python
class A:
    @property
    def p(self) -> int: return 1
    def m(self) -> int: return 1

class C:
    @property
    def p(self) -> int: return 2
    def m(self) -> int: return 2

def b_meth[T: A | C](x: T) -> int:   return x.m()   # invalid-argument-type, x2, this issue
def b_prop[T: A | C](x: T) -> int:   return x.p     # invalid-attribute-access, new in 0.0.70
def c_prop[T: (A, C)](x: T) -> int:  return x.p     # same, constrained
def u_prop(x: A | C) -> int:         return x.p     # ok, plain union
def s_prop[T: A](x: T) -> int:       return x.p     # ok, single-class bound
```

```text
error[invalid-attribute-access]: Invalid access to descriptor attribute `p` on type `T@b_prop`:
Expected `A`, found `T@b_prop`
```

| | mypy 1.19.1 | pyright 1.1.411 | ty 0.0.69 | ty 0.0.70 | ty 0.0.71 |
|---|---|---|---|---|---|
| method | ok | ok | error | error | error |
| property | ok | ok | **ok** | **error** | **error** |

I tried to get Claude to understand why I'm getting these errors, not sure how accurate the analysis is:

> The property half arrived with https://github.com/astral-sh/ruff/pull/27400,  Before it, a failed `__get__` call kept its return type and lost its error, so the wrong receiver was already there and merely invisible, then https://github.com/astral-sh/ruff/pull/27644, in 0.0.71, looks members up on each bound or constraint "while preserving the original receiver for descriptors and `Self`", so this survives it. Binding each member to an instance of that member, as proposed here, would cover descriptors too.

I hit it under helpers generic over two dataframe types that return whichever they were given, common in pandas and polars code:

```python
import pandas as pd

def first_non_empty[T: pd.DataFrame | pd.Series](*frames: T) -> T | None:
    return next((f for f in frames if not f.empty), None)   # error since 0.0.70
```

`.empty`, `.size`, `.shape` and `.index` are all properties and now error, `len(x)` is accepted.

**Comment by carljm** [comment]
Yes, I think we recently stopped silently dropping "bad `__get__` call" errors and unfortunately that exposed this bug more. We aim to get this fixed soon, sorry about that.
